### ISSUE TITLE
Cilium is not ready yet: connectivity health is failing

### ISSUE THREAD

**Issue by nebril** [cilium/cilium#11582]
We are hitting this across versions and tests, example output is

```
cilium pre-flight checks failed
Expected
    <*errors.errorString | 0xc0017b6e90>: {
        s: "Cilium validation failed: 4m0s timeout expired: Last polled error: connectivity health is failing: cilium-agent 'cilium-59fv4': Only 2/1 nodes appeared in cilium-health status. nodes = '[k8s2]'",
    }
to be nil
/home/jenkins/workspace/Ginkgo-CI-Tests-k8s1.18-Pipeline/k8s-1.18-gopath/src/github.com/cilium/cilium/test/helpers/manifest.go:297
```

Example failures:
https://jenkins.cilium.io/job/Ginkgo-CI-Tests-k8s1.18-Pipeline/368/testReport/Suite-k8s-1/18/K8sDatapathConfig_ManagedEtcd_Check_connectivity_with_managed_etcd/
https://jenkins.cilium.io/job/Ginkgo-CI-Tests-Pipeline-GKE/1180/testReport/junit/Suite-k8s-1/14/K8sConformance_Portmap_Chaining_Check_connectivity_check_compliance_with_portmap_chaining/

On GKE builds it usually correlates with multiple other test failures of similar nature.

[83230dcd_K8sConformance_Portmap_Chaining_Check_connectivity-check_compliance_with_portmap_chaining.zip](https://github.com/cilium/cilium/files/4644774/83230dcd_K8sConformance_Portmap_Chaining_Check_connectivity-check_compliance_with_portmap_chaining.zip)

**Comment by errordeveloper** [comment]
I am a bit confused about the error `Put \"http://%2Fvar%2Frun%2Fcilium%2Fhealth.sock/v1beta/status/probe\": context deadline exceeded`

Are we using HTTP over UNIX domain sockets, à la Docker?

I can see @tklauser committed d613dea5d52494da703e72ef0011e2849f7f19ce, which changes the probes that pods use, I am assuming it's not related, but I wonder - should these test switch to the new mechanism?

**Comment by michi-covalent** [comment]
i'm kind of suspecting a2ff935e4ca26a0fcc8359259447d4ff1414b8f7 looks like this was actually doing something to k8s nodes.

**Comment by glibsm** [comment]
can confirm, has been trying to figure this out the whole day locally

**Comment by tgraf** [comment]
Related to https://github.com/cilium/cilium/pull/11583?

**Comment by tklauser** [comment]
> I can see @tklauser committed d613dea, which changes the probes that pods use, I am assuming it's not related, but I wonder - should these tests switch to the new mechanism?

d613dea is using HTTP over TCP sockets on localhost, not unix domain sockets and also it's just for `cilium-agent` health checks. The error indicates something is wrong on the `cilium-health` UDS which is a separate status API. So I don't think it should be related.

**Comment by pchaigno** [comment]
> Related to #11583?

I don't think it's related as the host endpoint is in `ready` state on both nodes in both cases reported in the OP.

**Comment by tgraf** [comment]
We debugged this further in the Hubble meeting today:

Node to node connectivity is reported fine, connectivity to remote health endpoint fails:
```
# cilium-health status
Probe time:   2020-05-19T17:04:44Z
Nodes:
  gke-tgraf-test3-default-pool-537ea074-3w82 (localhost):
    Host connectivity to 10.164.0.68:
      ICMP to stack:   OK, RTT=234.344µs
      HTTP to agent:   OK, RTT=198.589µs
    Endpoint connectivity to 10.0.1.210:
      ICMP to stack:   OK, RTT=210.538µs
      HTTP to agent:   OK, RTT=385.375µs
  gke-tgraf-test3-default-pool-537ea074-hqsc:
    Host connectivity to 10.164.0.75:
      ICMP to stack:   OK, RTT=1.142154ms
      HTTP to agent:   OK, RTT=1.609108ms
    Endpoint connectivity to 10.0.0.249:
      ICMP to stack:   Connection timed out
      HTTP to agent:   Get "http://10.0.0.249:4240/hello": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

The port is reachable from the local node:
```
curl --fail -v 10.0.0.249:4240/hello
*   Trying 10.0.0.249:4240...
* TCP_NODELAY set
* Connected to 10.0.0.249 (10.0.0.249) port 4240 (#0)
> GET /hello HTTP/1.1
> Host: 10.0.0.249:4240
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Tue, 19 May 2020 17:07:46 GMT
< Content-Length: 0
<
* Connection #0 to host 10.0.0.249 left intact
```

The port is not reachable from a remote node:
```
curl --fail -v 10.0.0.249:4240/hello
*   Trying 10.0.0.249:4240...
* TCP_NODELAY set
```

Looking at the IP, it seems that the IP is not routable:
```
k get nodes -o yaml | grep -i cidr
      io.cilium.network.ipv4-pod-cidr: 10.0.1.0/24
    podCIDR: 10.4.0.0/24
      io.cilium.network.ipv4-pod-cidr: 10.0.0.0/24
    podCIDR: 10.4.1.0/24
```

The reason for this is:
```
  ipam: cluster-pool
```

**Comment by tgraf** [comment]
A fix for this is already queued up in this PR:
https://github.com/cilium/cilium/pull/11545/commits/d0e0fae4611b45065b9c439dc20097812f764e32

**Comment by pchaigno** [comment]
I think we're mixing too many flakes in this issue, so I'll close and have opened #11690 for the remaining `K8sDatapathConfig Etcd Check connectivity` flake.
- The GKE flake has been fixed in #11545.
- The test `K8sDatapathConfig ManagedEtcd Check connectivity with managed etcd` doesn't exist anymore and has been replaced by `K8sDatapathConfig Etcd Check connectivity`. It's not clear whether the last flake is related the same as we had on the old test.
